package/common_init/overlay/etc/init.d/S99reset_boot_counter

```diff
@@ -10,24 +10,38 @@ source /etc/init.d/logging.sh
 
 source /etc/init.d/common.sh
 
-upgrade_uboot_environment()
+initialize_uboot_environment()
 {
+  fw_setenv env_version_major 0
+  fw_setenv env_version_minor 1
+  fw_setenv env_version_minor_minor 0
   fw_setenv pre_boot_counter 0
   fw_setenv post_boot_counter 0
-  fw_setenv increment_boot_counter "setexpr pre_boot_counter \${pre_boot_counter} + 1; saveenv"
-  fw_setenv bootcmd "run increment_boot_counter; run fpgaload; run sdboot; run netboot;"
+  fw_setenv increment_boot_counter "setexpr pre_boot_counter \${pre_boot_counter} + 1; saveenv;"
+  fw_setenv pre_load_tasks "run increment_boot_counter;"
+  fw_setenv bootcmd "run pre_load_tasks; run fpgaload; run sdboot; run netboot;"
 }
 
-if fw_printenv pre_boot_counter; then
-  pre_boot_counter=$(fw_printenv -n pre_boot_counter)
-  post_boot_counter=$(fw_printenv -n post_boot_counter)
-  if [[ $pre_boot_counter != 1 || $post_boot_counter != 1 ]]; then
-    loge --sbp "Unexpected boot counter values! pre_boot_counter: ${pre_boot_counter} post_boot_counter ${post_boot_counter}"
+if fw_printenv env_version_major; then
+  env_version_major=$(fw_printenv -n env_version_major)
+  env_version_minor=$(fw_printenv -n env_version_minor)
+  env_version_minor_minor=$(fw_printenv -n env_version_minor_minor)
+
+  if [[ $env_version_major -ge 0 && $env_version_minor -ge 1 && $env_version_minor_minor -ge 0 ]]; then
+    pre_boot_counter=$(fw_printenv -n pre_boot_counter)
+    post_boot_counter=$(fw_printenv -n post_boot_counter)
+    if [[ $pre_boot_counter != 1 || $post_boot_counter != 1 ]]; then
+      if [[ $pre_boot_counter == 0 ]]; then
+        logw --sbp "pre_boot_counter = 0; Have you upgraded dev image?"
+      else
+        loge --sbp "Unexpected boot counter values! pre_boot_counter: ${pre_boot_counter} post_boot_counter ${post_boot_counter}"
+      fi;
+    fi;
+    logi --sbp "Resetting boot counters... pre_boot_counter: ${pre_boot_counter} post_boot_counter ${post_boot_counter}"
+    fw_setenv pre_boot_counter 0
+    fw_setenv post_boot_counter 0
   fi;
```

Comment:
default `else` statement maybe? not sure if how we'll handle multiple versions eventually

Reply:
My idea here was to segment off every version's additional functionality within version `if` block checks.

In this case, all versions `0.1.0` and above should have `pre/post_boot_counter` variables
Suppose we added a new variable, `new_boot_counter`, release would tick to `0.1.1`; code touching that counter would go inside a new if block

Series of `if` statements instead of `else if` allows to reuses previous version functionality; which i don't expect should change much between versions.